### PR TITLE
add tokens transferred spell

### PR DESCRIPTION
**Pull request by AnonJon**
Add new tokens transferred spell to access token information.

### CODE REVIEW

File: models/chainlink/arbitrum/chainlink_arbitrum_ccip_tokens_transferred.sql [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+    {{ref('chainlink_arbitrum_ccip_tokens_transferred_logs')}} le
+    INNER JOIN {{ref('chainlink_arbitrum_ccip_send_requested')}} sr ON le.tx_hash = sr.tx_hash
+    INNER JOIN  {{ source('erc20_arbitrum', 'evt_transfer') }} te ON le.contract_address = te.to
+    AND te.evt_tx_hash = le.tx_hash
+    LEFT JOIN {{ source('tokens', 'erc20') }} t ON te.contract_address = t.contract_address
+    LEFT JOIN {{ source('prices', 'usd') }} p ON te.contract_address = p.contract_address
+    AND p.minute = DATE_TRUNC ('minute', le.block_time)
+    {% if is_incremental() %}
+        WHERE {{ incremental_predicate('block_time') }}
+    {% endif %}
```
Comment (jeff-dude) [inline models/chainlink/arbitrum/chainlink_arbitrum_ccip_tokens_transferred.sql:40]: any time you are filtering on a table incrementally, you should apply the same filter in join conditions to other tables being brought in. please apply incremental filter on all joins which contain data in time-series format (i.e. prices, evt_transfer, other chainlink refs, etc.)

**Review by jeff-dude — CHANGES_REQUESTED** [review]
@AnonJon i'd love to use this PR as an opportunity to introduce dbt macros to CCIP spells. since you're introducing a new spell, and the code looks repeatable across evm chains, it should fit perfectly.

please see here for initial context on how to approach:
https://github.com/duneanalytics/spellbook/blob/main/docs/macros/macro_overview.md

we can help you along the way if you happen to get stuck on anything.

**Comment by AnonJon** [comment]
@jeff-dude I implemented a macro for the latest spell. Tests are taking longer than expected so let me know if ive set up everything correctly.

File: macros/models/_project/chainlink/chainlink_ccip_macro.sql [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+        le.total_tokens / POWER(
+            CAST(10 AS double precision),
+            CAST(t.decimals AS double precision)
+        )
+    ), 0) AS token_price
+FROM
+    {{ref('chainlink_' ~ blockchain ~ '_ccip_tokens_transferred_logs')}} le
+    INNER JOIN {{ref('chainlink_' ~ blockchain ~ '_ccip_send_requested')}} sr ON le.tx_hash = sr.tx_hash
+    {% if is_incremental() %}
+        WHERE {{ incremental_predicate('block_time') }}
```
Comment (jeff-dude) [inline macros/models/_project/chainlink/chainlink_ccip_macro.sql:28]:
```suggestion
        AND {{ incremental_predicate('sr.block_time') }}
```

File: macros/models/_project/chainlink/chainlink_ccip_macro.sql [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    COALESCE(t.symbol, 'UNKNOWN') AS token_symbol,
+    COALESCE(p.price * (
+        le.total_tokens / POWER(
+            CAST(10 AS double precision),
+            CAST(t.decimals AS double precision)
+        )
+    ), 0) AS token_price
+FROM
+    {{ref('chainlink_' ~ blockchain ~ '_ccip_tokens_transferred_logs')}} le
+    INNER JOIN {{ref('chainlink_' ~ blockchain ~ '_ccip_send_requested')}} sr ON le.tx_hash = sr.tx_hash
```
Comment (jeff-dude) [inline macros/models/_project/chainlink/chainlink_ccip_macro.sql:26]: are there more join conditions we can add here to improve performance?
`block_number`? `block_date`? `block_time`? `evt_index`?

Comment (jeff-dude) [reply macros/models/_project/chainlink/chainlink_ccip_macro.sql:26]: joining only on varbinary hash values is a bit slow

File: macros/models/_project/chainlink/chainlink_ccip_macro.sql [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+    {% if is_incremental() %}
+        WHERE {{ incremental_predicate('block_time') }}
+    {% endif %}
+    INNER JOIN  {{ source('erc20_' ~ blockchain, 'evt_transfer') }} te ON le.contract_address = te.to
+    AND te.evt_tx_hash = le.tx_hash
```
Comment (jeff-dude) [inline macros/models/_project/chainlink/chainlink_ccip_macro.sql:31]: similar question here on potential for more join conditions

File: macros/models/_project/chainlink/chainlink_ccip_macro.sql [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+    {% if is_incremental() %}
+        WHERE {{ incremental_predicate('block_time') }}
```
Comment (jeff-dude) [inline macros/models/_project/chainlink/chainlink_ccip_macro.sql:33]:
```suggestion
        AND {{ incremental_predicate('te.block_time') }}
```

File: macros/models/_project/chainlink/chainlink_ccip_macro.sql [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+    {% endif %}
+    LEFT JOIN {{ source('tokens', 'erc20') }} t ON te.contract_address = t.contract_address
```
Comment (jeff-dude) [inline macros/models/_project/chainlink/chainlink_ccip_macro.sql:35]: and blockchain = blockchain?

File: macros/models/_project/chainlink/chainlink_ccip_macro.sql [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+    {% if is_incremental() %}
+        AND {{ incremental_predicate('block_time') }}
+    {% endif %}
```
Comment (jeff-dude) [inline macros/models/_project/chainlink/chainlink_ccip_macro.sql:38]:
```suggestion

```

Comment (jeff-dude) [reply macros/models/_project/chainlink/chainlink_ccip_macro.sql:38]: token metadata doesn't need to be filtered incrementally, it's just one row per address/chain

File: macros/models/_project/chainlink/chainlink_ccip_macro.sql [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+    LEFT JOIN {{ source('prices', 'usd') }} p ON te.contract_address = p.contract_address
+    AND p.minute = DATE_TRUNC ('minute', le.block_time)
```
Comment (jeff-dude) [inline macros/models/_project/chainlink/chainlink_ccip_macro.sql:40]: and blockchain = blockchain?

File: macros/models/_project/chainlink/chainlink_ccip_macro.sql [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+    {% if is_incremental() %}
+        WHERE {{ incremental_predicate('block_time') }}
```
Comment (jeff-dude) [inline macros/models/_project/chainlink/chainlink_ccip_macro.sql:42]:
```suggestion
        AND {{ incremental_predicate('p.minute') }}
```

File: models/chainlink/chainlink_ccip_tokens_transferred.sql [hunk 10]
```diff
@@ -0,0 +1,39 @@
+{{
+  config(
+
```
Comment (jeff-dude) [inline models/chainlink/chainlink_ccip_tokens_transferred.sql:3]: add schema property here plz

**Review by jeff-dude — CHANGES_REQUESTED** [review]

**Review by jeff-dude — APPROVED** [review]
code is looking pretty good to me -- hope you find this approach useful 🙏  looking forward to seeing this more in the chainlink spells.

the performance is still a bit slow, but looks like it's a lot of data and joins to transfers tables.

**Pull request merged**